pkg/alertmanager/distributor_test.go

```diff
@@ -0,0 +1,297 @@
+		func(_ interface{}) (interface{}, bool, error) {
+			return &ring.Desc{
+				Ingesters: amDescs,
+			}, true, nil
+		},
+	)
+	require.NoError(t, err)
+
+	amRing, err := ring.New(ring.Config{
+		KVStore: kv.Config{
+			Mock: kvStore,
+		},
+		HeartbeatTimeout:  60 * time.Minute,
+		ReplicationFactor: replicationFactor,
+	}, RingNameForServer, RingKey, nil)
+	require.NoError(t, err)
+	require.NoError(t, services.StartAndAwaitRunning(context.Background(), amRing))
+	test.Poll(t, time.Second, numAM, func() interface{} {
+		return amRing.InstancesCount()
+	})
+
+	cfg := &MultitenantAlertmanagerConfig{}
+	flagext.DefaultValues(cfg)
+
+	d, err := NewDistributor(cfg, amRing, newMockAlertmanagerClientFactory(amByAddr), util_log.Logger, prometheus.NewRegistry())
+	require.NoError(t, err)
+	require.NoError(t, services.StartAndAwaitRunning(context.Background(), d))
+
+	return d, ams, func() {
+		require.NoError(t, services.StopAndAwaitTerminated(context.Background(), d))
+	}
+}
+
+type mockAlertmanager struct {
+	alertmanagerpb.AlertmanagerClient
+	grpc_health_v1.HealthClient
+	t *testing.T
+	// receivedRequests is map of route -> statusCode -> number of requests.
+	receivedRequests map[string]map[int]int
+	mtx              sync.Mutex
+	myAddr           string
+	happy            bool
+}
+
+func newMockAlertmanager(t *testing.T, idx int, happy bool) *mockAlertmanager {
+	return &mockAlertmanager{
+		t:                t,
+		receivedRequests: make(map[string]map[int]int),
+		myAddr:           fmt.Sprintf("127.0.0.1:%05d", 10000+idx),
+		happy:            happy,
+	}
+}
+
+func (am *mockAlertmanager) HandleRequest(_ context.Context, in *httpgrpc.HTTPRequest, _ ...grpc.CallOption) (*httpgrpc.HTTPResponse, error) {
+	am.mtx.Lock()
+	defer am.mtx.Unlock()
+
+	u, err := url.Parse(in.Url)
+	require.NoError(am.t, err)
+	path := u.Path
+	m, ok := am.receivedRequests[path]
+	if !ok {
+		m = make(map[int]int)
+		am.receivedRequests[path] = m
+	}
+
+	if am.happy {
+		m[http.StatusOK]++
+		return &httpgrpc.HTTPResponse{
+			Code: http.StatusOK,
+			Headers: []*httpgrpc.Header{
+				{
+					Key:    "Content-Type",
+					Values: []string{"it-is-ok"},
+				}, {
+					Key:    "X-Content-Type-Options",
+					Values: []string{"ok-option-1", "ok-option-2"},
+				},
+			},
+		}, nil
+	}
+
+	m[http.StatusInternalServerError]++
+	return nil, httpgrpc.ErrorFromHTTPResponse(&httpgrpc.HTTPResponse{
+		Code: http.StatusInternalServerError,
+		Headers: []*httpgrpc.Header{
+			{
+				Key:    "Content-Type",
+				Values: []string{"it-is-not-ok"},
+			}, {
+				Key:    "X-Content-Type-Options",
+				Values: []string{"not-ok-option-1", "not-ok-option-2"},
+			},
+		},
+	})
+}
+
+func (am *mockAlertmanager) requestsCount(route string) int {
+	am.mtx.Lock()
+	defer am.mtx.Unlock()
+
+	routesReceived := len(am.receivedRequests)
+	if routesReceived == 0 {
+		// This AM did not receive any requests.
+		return 0
+	}
+
+	require.Equal(am.t, 1, routesReceived)
```

Comment:
A generic mock should assert on the number of requests received. I would remove this.